src/Grpc.AspNetCore.Server.Reflection/GrpcReflectionServiceExtensions.cs

```diff
@@ -41,9 +45,71 @@ public static IServiceCollection AddGrpcReflection(this IServiceCollection servi
                 throw new ArgumentNullException(nameof(services));
             }
 
-            services.TryAddSingleton<IGrpcServiceActivator<ReflectionServiceImpl>, ReflectionGrpcServiceActivator>();
+            // ReflectionServiceImpl is designed to be a singleton
+            // Explicitly register creating it in DI using descriptors calculated from gRPC endpoints in the app
+            services.TryAddSingleton<ReflectionServiceImpl>(serviceProvider =>
+            {
+                var logger = serviceProvider.GetRequiredService<ILoggerFactory>().CreateLogger(nameof(GrpcReflectionServiceExtensions));
+                var endpointDataSource = serviceProvider.GetRequiredService<EndpointDataSource>();
+
+                var grpcEndpointMetadata = endpointDataSource.Endpoints
+                    .Select(ep => ep.Metadata.GetMetadata<GrpcMethodMetadata>())
+                    .Where(m => m != null)
+                    .ToList();
+
+                var serviceTypes = grpcEndpointMetadata.Select(m => m.ServiceType).Distinct().ToList();
+
+                var serviceDescriptors = new List<Google.Protobuf.Reflection.ServiceDescriptor>();
+
+                foreach (var serviceType in serviceTypes)
+                {
+                    var baseType = GetServiceBaseType(serviceType);
+                    var definitionType = baseType?.DeclaringType;
+
+                    var descriptorPropertyInfo = definitionType?.GetProperty("Descriptor", BindingFlags.Public | BindingFlags.Static);
+                    if (descriptorPropertyInfo != null)
+                    {
+                        var serviceDescriptor = descriptorPropertyInfo.GetValue(null) as Google.Protobuf.Reflection.ServiceDescriptor;
+                        if (serviceDescriptor != null)
+                        {
+                            serviceDescriptors.Add(serviceDescriptor);
+                            continue;
+                        }
+                    }
+
+                    Log.ServiceDescriptorNotResolved(logger, serviceType);
+                }
+
+                return new ReflectionServiceImpl(serviceDescriptors);
```

Comment:
Any reason not to add a ReflectionServiceImpl ctor that takes an ILoggerFactory and a EndpointDataSource as parameters and does all this stuff in the ctor? Is this closing over any variables? Putting this in a ctor seems better to me than putting this logic in a service factory lambda.

Reply:
Yes this is kinda ugly.

Reply:
`ReflectionServiceImpl` is from an external package.